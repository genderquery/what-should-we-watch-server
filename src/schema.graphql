type Query {
  movie(id: ID!): Movie
}

type Movie {
  id: ID!
  title: String!
  releaseDate: String!
  overview: String!
}
