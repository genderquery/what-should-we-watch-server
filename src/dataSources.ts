import { RequestOptions, RESTDataSource } from "apollo-datasource-rest";

export class TmdbDataSource extends RESTDataSource {
  private apiToken: string;

  constructor(apiToken: string) {
    super();
    this.apiToken = apiToken;
    this.baseURL = "https://api.themoviedb.org/";
  }

  willSendRequest(request: RequestOptions) {
    request.headers.set("Authorization", `Bearer ${this.apiToken}`);
  }

  static movieReducer(source: any) {
    return {
      id: source.id,
      title: source.title,
      releaseDate: source.release_date,
      overview: source.overview,
    };
  }

  async getMovieById(id: number) {
    const result = await this.get(`3/movie/${id}`);
    return TmdbDataSource.movieReducer(result);
  }
}

export default () => ({
  tmdb: new TmdbDataSource(process.env.TMDB_API_TOKEN!),
});
