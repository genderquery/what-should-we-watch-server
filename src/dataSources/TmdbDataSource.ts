import { RequestOptions, RESTDataSource } from "apollo-datasource-rest";

export class TmdbDataSource<Context> extends RESTDataSource<Context> {
  private apiToken: string;

  constructor(apiToken: string) {
    super();
    this.apiToken = apiToken;
    this.baseURL = "https://api.themoviedb.org/";
  }

  willSendRequest(request: RequestOptions): void {
    request.headers.set("Authorization", `Bearer ${this.apiToken}`);
  }

  static movieReducer(source: any) {
    return {
      id: source.id,
      title: source.title,
      releaseDate: source.release_date,
      overview: source.overview,
    };
  }

  async getMovieById(id: string) {
    const result = await this.get(`3/movie/${id}`);
    return TmdbDataSource.movieReducer(result);
  }
}

export default TmdbDataSource;
