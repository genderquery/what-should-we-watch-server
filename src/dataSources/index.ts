import { DataSource } from "apollo-datasource";
import { TmdbDataSource } from "./TmdbDataSource";

export interface DataSources<Context>
  extends Record<string, DataSource<Context>> {
  tmdb: TmdbDataSource<Context>;
}

const dataSources = () => ({
  tmdb: new TmdbDataSource(process.env.TMDB_API_TOKEN!),
});

export { TmdbDataSource };
export default dataSources;
