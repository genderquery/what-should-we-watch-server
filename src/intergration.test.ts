import { createTestClient } from "apollo-server-testing";
import { ApolloServer, gql } from "apollo-server-express";
import { mocked } from "ts-jest/utils";
import schema from "./schema";
import { TmdbDataSource } from "./dataSources";

jest.mock("./dataSources");

const createTestServer = () => {
  const tmdb = mocked(new TmdbDataSource(""), true);
  const server = new ApolloServer({
    schema,
    dataSources: () => ({ tmdb }),
  });
  return { dataSources: { tmdb }, server };
};

const movieDetailQuery = gql`
  query MovieDetail($id: ID!) {
    movie(id: $id) {
      id
      title
      releaseDate
      overview
    }
  }
`;

test("query movie detail", async () => {
  const { dataSources, server } = createTestServer();
  dataSources.tmdb.getMovieById.mockResolvedValue(mockMovie);
  const { query } = createTestClient(server);
  const result = await query({
    query: movieDetailQuery,
    variables: {
      id: 603,
    },
  });
  expect(result).toMatchSnapshot();
});

const mockMovie = {
  id: 603,
  overview:
    "Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.",
  releaseDate: "1999-03-30",
  title: "The Matrix",
};
