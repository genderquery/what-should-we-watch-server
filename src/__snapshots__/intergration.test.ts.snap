// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query movie detail 1`] = `
Object {
  "data": Object {
    "movie": Object {
      "id": "603",
      "overview": "Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.",
      "releaseDate": "1999-03-30",
      "title": "The Matrix",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
